code/game/machinery/pipe/construction.dm

```diff
@@ -208,40 +229,51 @@ GLOBAL_LIST_INIT(pipeID2State, list(
 /obj/item/pipe/attack_self(mob/user)
 	return rotate()
 
+/obj/item/pipe/proc/get_pipe_cache(type)
+	var/static/list/obj/machinery/atmospherics/check_cache
+	if(!islist(check_cache))
+		check_cache = list()
+	if(!check_cache[type])
+		check_cache[type] = new type
+	return check_cache[type]
+
 /obj/item/pipe/attackby(obj/item/W, mob/user, params)
 	if (!istype(W, /obj/item/wrench))
 		return ..()
-	if (!isturf(src.loc))
-		return 1
-	
+	if (!isturf(loc))
+		return TRUE
 	add_fingerprint(user)
 
 	fixdir()
 	if(pipe_type in list(PIPE_GAS_MIXER, PIPE_GAS_FILTER))
 		setDir(unflip(dir))
 
-	var/obj/machinery/atmospherics/A = new pipe_type(src.loc)
-	A.setDir(src.dir)
-	A.SetInitDirections()
+	var/obj/machinery/atmospherics/fakeA = get_pipe_cache(pipe_type)
 
-	for(var/obj/machinery/atmospherics/M in src.loc)
-		if(M == A) //we don't want to check to see if it interferes with itself
+	for(var/obj/machinery/atmospherics/M in loc)
+		if((M.pipe_flags & PIPING_ONE_PER_TURF) && (fakeA.pipe_flags & PIPING_ONE_PER_TURF))	//Only one dense/requires density object per tile, eg connectors/cryo/heater/coolers.
+			to_chat(user, "<span class='warning'>Something is hogging the tile!</span>")
+			return TRUE
+		if((M.piping_layer != piping_layer) && !((M.pipe_flags & PIPING_ALL_LAYER) || (pipe_type == PIPE_LAYER_MANIFOLD)))
 			continue
-		if(M.GetInitDirections() & A.GetInitDirections())	// matches at least one direction on either type of pipe
+		if(M.GetInitDirections() & fakeA.GetInitDirections())	// matches at least one direction on either type of pipe

```

Comment:
Are you only populating your cache with one dir?

Reply:
oshit